### PR TITLE
Add C API function that returns all parameter names with their aliases

### LINKED ISSUE #2633: Add C API function that returns all parameter names with their aliases
Refer to https://github.com/microsoft/LightGBM/pull/2594#issuecomment-563356223.

> C API function which returns all param names with their aliases? For example, it can be a string format like `param1:param1_alias1,param1_alias2;param2:param2_alias1;param3;param4:...`
It will allow to deduplicate aliases handling code in language wrappers and create param aliases handling mechanism by simple string parsing got from C API.

> it could be done by the config code automatically generating.
As for the objective and metric, we may need to add them manually.

### PR DESCRIPTION
**Pull request by StrikerRUS**
Fixed #2633

### CODE REVIEW

File: R-package/R/aliases.R
Comment (StrikerRUS) [inline R-package/R/aliases.R:48]: @jameslamb Could you please help with R code here? The failing test is
```
  == Failed ======================================================================
  -- 1. Failure (test_parameters.R:60:3): .PARAMETER_ALIASES() returns a named lis
  all(sapply(param_aliases, is.character)) is not TRUE
  
  `actual`:   FALSE
  `expected`: TRUE 
```
and the problem is in keys with empty lists:
```R
aliases <- jsonlite::fromJSON("{\"task\": [\"task_type\", \"test\"], \"config\": [\"config_file\"], \"boosting\": [\"boosting_type\", \"boost\"], \"pre_filter\": []}")

for (alias in names(aliases)) {
    aliases[[alias]] <- c(aliases[[alias]], alias)
}

all(sapply(aliases, is.character))
length(unique(names(aliases))) == length(aliases)
print(aliases)
```
```
[1] FALSE
[1] TRUE
$task
[1] "task_type" "test"      "task"     

$config
[1] "config_file" "config"     

$boosting
[1] "boosting_type" "boost"         "boosting"     

$pre_filter
$pre_filter[[1]]
[1] "pre_filter"
```
Note `pre_filter` in the example above.

Comment (StrikerRUS) [reply R-package/R/aliases.R]: Looks like I've managed to fix this, but will be happy to see a more elegant way 🙂

**Comment by StrikerRUS** [comment]
OK, I believe this PR is ready for reviews.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [inline python-package/lightgbm/basic.py:332]:
```suggestion
    def _get_all_param_aliases() -> Dict[str, Set[str]]:
```

Could you please add a return type hint here, to help `mypy` possibly catch issues? (per #3867)

**Review by jameslamb — CHANGES_REQUESTED** [review]
This is awesome! I totally support this addition to the API. I think it'll help to avoid some maintenance burden in language wrappers and will eliminate some possible bugs. Great work 🤩 

Please see a few small comments I left.

File: include/LightGBM/config.h
Comment (StrikerRUS) [reply include/LightGBM/config.h]: I have already tried to make this function `const` but all my attempts failed.
Pushed your suggestion in a0745f2e6085743c56b110267b8c98e949605a75.

Comment (StrikerRUS) [reply include/LightGBM/config.h]: Doesn't work:
```
/__w/1/s/src/io/config_auto.cpp:759:13: error: prototype for ‘std::string LightGBM::Config::DumpAliases()’ does not match any in class ‘LightGBM::Config’
 std::string Config::DumpAliases() {
             ^
In file included from /__w/1/s/src/io/config_auto.cpp:8:0:
/__w/1/s/include/LightGBM/config.h:1044:28: error: candidate is: static const string LightGBM::Config::DumpAliases()
   static const std::string DumpAliases();
                            ^
```

Comment (jameslamb) [reply include/LightGBM/config.h]: Oh! You'd have to change this both here in the header file and in the implementation file (in this case `config_auto.cpp`). These errors are the compiler complaining about those not matching.

I pulled this branch and tested today, works fine. See the diff on this branch on my fork: https://github.com/jameslamb/LightGBM/commit/df641a8acf7205e56b415babf0d95fd6f0b49d91

Comment (StrikerRUS) [reply include/LightGBM/config.h]: Thanks a lot for the tip! Added in 9a73cb4de65004b9d5740ebf4f19997f1b84f986.

**Comment by StrikerRUS** [comment]
@jameslamb Thanks a lot for your review comments! I think I've addressed them all in the latest commit. Please take another look when have time.

**Review by jameslamb — APPROVED** [review]
Looks great, thanks for the changes, especially the new R test!

Since this touched `lightgbm_R.cpp`, I just did a `valgrind` run (sorry for not thinking of that sooner). https://github.com/microsoft/LightGBM/pull/4829#issuecomment-985019519

As soon as that passes, I'm comfortable with merging this.

**Comment by jameslamb** [comment]
@shiyu1994 do you support this addition to `c_api.cpp`? I think changes to that file should probably have two approvals, just since it is the core public API of LightGBM.

**Review by shiyu1994 — APPROVED** [review]
@StrikerRUS @jameslamb Thank you for working on this! I've checked the C API, and the changes LGTM.

**Pull request merged**
